ouroboros-network/test/Test/Mux.hs

```diff
@@ -390,61 +517,275 @@ prop_mux_starvation response0 response1 =
+                    , MonadMask m
+                    , MonadSay m
+                    , MonadST m
+                    , MonadSTM m
+                    , MonadThrow (STM m)
+                    , MonadTime m
+                    , MonadTimer m)
+                 => ArbitrarySDU
+                 -> m Property
+prop_demux_sdu a = do
+    r <- run a
+    return $ tabulate "SDU type" [stateLabel a] $
+             tabulate "SDU Violation " [violationLabel a] r
+  where
+    run (ArbitraryValidSDU sdu state (Just Mx.MuxIngressQueueOverRun)) = do
+        stopVar <- newEmptyTMVarM
 
-    let sduLen = 1260
-        request  = DummyPayload $ BL.replicate 4 0xa
-        response = request
+        let server_mp = Mx.MiniProtocolDescription Nothing (Just $ serverRsp stopVar)
+        -- To trigger MuxIngressQueueOverRun we use a special test protocol with
+        -- an ingress queue which is less than 0xffff so that it can be triggered by a
+        -- single segment.
+        let server_small ChainSyncSmall = server_mp
+
+        (client_w, said) <- plainServer server_small
+        setup state client_w
+
+        writeSdu client_w $! unDummyPayload sdu
+
+        atomically $! putTMVar stopVar $ unDummyPayload sdu
 
-    (_, _, server_mp) <- setupMiniReqRsp (return ()) endMpsVar request response
-    let server_mps ChainSync1 = server_mp
+        res <- wait said
+        case res of
+            Just e  ->
+                case fromException e of
+                    Just me -> return $ Mx.errorType me === Mx.MuxIngressQueueOverRun
+                    Nothing -> return $ property False
+            Nothing -> return $ property False
+
+    run (ArbitraryValidSDU sdu state err_m) = do
+        stopVar <- newEmptyTMVarM
+
+        let server_mp = Mx.MiniProtocolDescription Nothing (Just $ serverRsp stopVar)
+        let server_std ChainSync1 = server_mp
+
+        (client_w, said) <- plainServer server_std
+
+        setup state client_w
+
+        atomically $! putTMVar stopVar $! unDummyPayload sdu
+        writeSdu client_w $ unDummyPayload sdu
+
+        res <- wait said
+        case res of
+            Just e  ->
+                case fromException e of
+                    Just me -> case err_m of
+                                    Just err -> return $ Mx.errorType me === err
+                                    Nothing  -> return $ property False
+                    Nothing -> return $ property False
+            Nothing -> return $ err_m === Nothing
+
+    run (ArbitraryInvalidSDU badSdu state err) = do
+        stopVar <- newEmptyTMVarM
+
+        let server_mp = Mx.MiniProtocolDescription Nothing (Just $ serverRsp stopVar)
+        let server_std ChainSync1 = server_mp
+
+        (client_w, said) <- plainServer server_std
+
+        setup state client_w
+        atomically $ writeTBQueue client_w $ BL.take (isRealLength badSdu) $ encodeInvalidMuxSDU badSdu
+        atomically $ putTMVar stopVar BL.empty
```

Comment:
Hmm, this means that the server will never read from the channel and still the mux will throw the exception, :thinking:

Reply:
Good catch! This is racy.
The `demux` thread will read the corrupt header and throw the exception so the test case will normally pass, but if the server thread is faster it will notice the empty string and exit successfully causing the testcase to fail.

Reply:
Have you tried to address this? Maybe at least document that this might happen, so it will be easier to address if this happens.

Reply:
Yes, sorry for not making myself clear. It is fixed, see https://github.com/input-output-hk/ouroboros-network/blob/1db7f4622fe90bb30a6fb5e51b3ffe2d4a9d2fc4/ouroboros-network/test/Test/Mux.hs#L619